pep-536.txt

```diff
@@ -0,0 +1,174 @@
+so this text requires familiarity with PEP 498.
+
+Terminology
+===========
+
+This text will refer to the existing grammar as “f-strings”,
+and the proposed one as “f-literals”.
+
+Furthermore, it will refer to the ``{}``-delimited expressions in
+f-literals/f-strings as “expression portions” and the static string content
+around them as “string portions”.
+
+Motivation
+==========
+
+The current implementation of f-strings in CPython relies on the existing
+string parsing machinery and a post processing of its tokens.  This results in
+several restrictions to the possible expressions usable within f-strings:
+
+#. It is impossible to use the quote character delimiting the f-string
+   within the expression portion::
+
+    >>> f'Magic wand: { bag['wand'] }'
+                                 ^
+    SyntaxError: invalid syntax
+
+#. A previously considered way around it would lead to escape sequences
+   in executed code and is prohibited in f-strings::
+
+    >>> f'Magic wand { bag[\'wand\'] } string'
+    SyntaxError: f-string expression portion cannot include a backslash
+
+#. Comments are forbidden even in multi-line f-strings::
+
+    >>> f'''A complex trick: {
+    ... bag['bag']  # recursive bags!
+    ... }'''
+    SyntaxError: f-string expression part cannot include '#'
+
+#. Expression portions need to wrap ``':'`` and ``'!'`` in braces::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
+
+These limitations are non-obvious and counter-intuitive unless the user is
+familiar with the f-literals’ implementation details.  Furthermore,
+they serve no purpose from a language user perspective and can be lifted
+by giving f-literals a regular grammar without exceptions
+and implementing it using dedicated parse code.
+
+.. note:: put preceding paragraph into “Rationale”?
+
+Rationale
+=========
+
+.. https://mail.python.org/pipermail/python-ideas/2016-August/041727.html
+
+As mentioned, a previous version of PEP 498 allowed escape sequences
+anywhere in f-strings, including as ways to encode the braces delimiting
+the expression portions and in their code.  They would be expanded before
+the code is parsed, which would have had several important ramifications:
+
+#. It would not be clear to human readers which portions are Expressions
+and which are strings.  Great material for an “obfuscated/underhanded
+Python challenge”
+#. Syntax highlighters are good in parsing nested grammar, but not
+in recognizing escape sequences.  ECMAScript 2016 (JavaScript) allows
+escape sequences in its identifiers [#]_ and the author knows of no
+syntax highlighter able to correctly highlight code making use of this.
+
+As a consequence, the expression portions would be harder to recognize
+with and without the aid of syntax highlighting.  With the new grammar,
+it is easy to extend syntax highlighters to correctly parse
+and display f-literals:
+
+.. raw:: html
```

Comment:
> I think it would be more interesting to actually include the diff of what you had to change to the syntax highlighter to make it work

good idea, i’ll add that part!

> than the actual output

seeing is believing! i really want to include that part for people who have only seen string-colored code parts until now instead of something like this!

![pep-0999-1](https://cloud.githubusercontent.com/assets/291575/21162114/fc2c09ca-c18b-11e6-985e-cb8974602362.png)

> I don't think we should change how python.org functions just for this one thing

actually @berkerpeksag was mistaken and we don’t have to change a thing. `generate_pep_pages.py` says:

> [Run this command AFTER normal RST -> HTML PEP transformation from the PEP
    repository has happened. This works on the HTML files created during that
    process..](https://github.com/python/pythondotorg/blob/0d99645d28b8e3935c0a5e83fd096a3ab59af70b/peps/management/commands/generate_pep_pages.py#L18-L20)

…and `pep2html.py` happily embeds my raw HTML